# Assignment-4

## 1. What assumptions are made about the attributes in Naïve Bayes method of classification and why? What is Laplacian correction and why it is necessary?

### Assumptions about the attributes in Naive Bayes method of classification:

#### 1. Independence Assumption:In Naive Bayes, we assume the features/attributes that go into the model is independent of each other. That is changing the value of one feature, does not directly influence or change the value of any of the other features/attributes used in the algorithm.  Naive Bayes assumes that the presence or absence of a particular feature is independent of the presence or absence of any other feature, given the class label. This is a strong assumption and may not always hold true in real-world data, hence the name "naive."

#### 2. Attribute Distribution Assumption: Naive Bayes often assumes a specific distribution of the data, such as a Gaussian distribution for continuous features or a multinomial distribution for discrete features. This assumption simplifies the computation of probabilities.

#### Laplacian correction is a technique used to handle the problem of zero probabilities in the Naive Bayes classifier. When we are calculating probabilities, if a particular attribute-value pair has not been observed in the training data for a given class, the probability estimate for that pair becomes zero. However, this can be problematic because it leads to a complete lack of flexibility in the model.

#### Laplacian correction is necessary because it addresses this issue by adding a small constant (usually 1) to the count of each possible attribute-value pair for each class. This ensures that no probability estimate is ever zero and prevents the model from being too rigid.

## 2. Take the iris data. Use Naïve Bayes algorithm to find the species with the following two observations. [4.5 3.0 5.6 2.1; 5.4 2.6 4.5 0.0] Show various steps involved and do as explained in the class

In [1]:
# importing necessary libraries
import numpy as np
import pandas as pd
from sklearn import datasets 
from sklearn.metrics import confusion_matrix 
from sklearn.model_selection import train_test_split 
import math

In [2]:
# loading the iris dataset 
iris = datasets.load_iris()
iris.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [3]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [4]:
X = iris.data
y =iris.target
print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


##### We get its encoding asabove, setosa:0, versicolor:1, virginica:2

In [5]:
iris.DESCR

'.. _iris_dataset:\n\nIris plants dataset\n--------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n                \n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)\n    ============== ==== =

In [6]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [7]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [8]:
pd.DataFrame(X).head()

0    1    2    3
0  5.1  3.5  1.4  0.2
1  4.9  3.0  1.4  0.2
2  4.7  3.2  1.3  0.2
3  4.6  3.1  1.5  0.2
4  5.0  3.6  1.4  0.2

In [9]:
(pd.DataFrame(X)).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       150 non-null    float64
 1   1       150 non-null    float64
 2   2       150 non-null    float64
 3   3       150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


In [10]:
pd.DataFrame(X).describe()

0           1           2           3
count  150.000000  150.000000  150.000000  150.000000
mean     5.843333    3.057333    3.758000    1.199333
std      0.828066    0.435866    1.765298    0.762238
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.350000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

In [11]:
df_x = pd.DataFrame(X)
df_x.columns = ['SL', 'SW', "PL","PW"]

In [12]:
df_x.describe()

SL          SW          PL          PW
count  150.000000  150.000000  150.000000  150.000000
mean     5.843333    3.057333    3.758000    1.199333
std      0.828066    0.435866    1.765298    0.762238
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.350000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

In [13]:
pd.DataFrame(X).isnull().sum()

0    0
1    0
2    0
3    0
dtype: int64

In [14]:
df_x = pd.DataFrame(X)
df_x.columns = ['SL', 'SW', "PL","PW"]
df_x.describe()

SL          SW          PL          PW
count  150.000000  150.000000  150.000000  150.000000
mean     5.843333    3.057333    3.758000    1.199333
std      0.828066    0.435866    1.765298    0.762238
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.350000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

In [15]:
#iris.data -> X_train
X_train=df_x
X_train

SL   SW   PL   PW
0    5.1  3.5  1.4  0.2
1    4.9  3.0  1.4  0.2
2    4.7  3.2  1.3  0.2
3    4.6  3.1  1.5  0.2
4    5.0  3.6  1.4  0.2
..   ...  ...  ...  ...
145  6.7  3.0  5.2  2.3
146  6.3  2.5  5.0  1.9
147  6.5  3.0  5.2  2.0
148  6.2  3.4  5.4  2.3
149  5.9  3.0  5.1  1.8

[150 rows x 4 columns]

In [16]:
# #iris.target -> y_train
y_train=y
y_train

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### Naïve Bayes calculates the probability of an event in the following steps:
#### 1.Calculate prior probability for given class labels
#### 2.Calculate conditional probability with each attributes for each class
#### 3.Multiply same class conditional probability
#### 4.Multiply prior probability with probability obtained in 3
#### 5.See which class has higher probability. Higher probability class belongs to given input set.

In [17]:
# 1.Calculate priori probability for given class labels
class_counts = pd.Series(y_train).value_counts()
prior_probabilities = class_counts / len(y_train)
prior_probabilities

0    0.333333
1    0.333333
2    0.333333
dtype: float64

#### prior probability of setosa is 0.333333 , prior probability of versicolor is 0.333333 and prior probability of virginica is 0.333333

In [18]:
# 2.Calculate conditional probability with each attributes for each class
# Calculate std and mean for each class for each feature
#Setosa
setosa =  X_train[y_train==0]
# Calculate mean and standard deviation for each feature
setosa_mean = setosa.mean()
print("Setosa mean\n",setosa_mean)
setosa_std = setosa.std()
print("Setosa Standard Deviation\n",setosa_std)

Setosa mean
 SL    5.006
SW    3.428
PL    1.462
PW    0.246
dtype: float64
Setosa Standard Deviation
 SL    0.352490
SW    0.379064
PL    0.173664
PW    0.105386
dtype: float64


In [19]:
#versicolor
versicolor =  X_train[y_train==1]
# Calculate mean and standard deviation for each feature
versicolor_mean = versicolor.mean()
print("Versicolor mean\n",versicolor_mean)
versicolor_std = versicolor.std()
print("Versicolor Standard Deviation\n",versicolor_std)

Versicolor mean
 SL    5.936
SW    2.770
PL    4.260
PW    1.326
dtype: float64
Versicolor Standard Deviation
 SL    0.516171
SW    0.313798
PL    0.469911
PW    0.197753
dtype: float64


In [20]:
#Virginica
virginica =  X_train[y_train==2]
# Calculate mean and standard deviation for each feature
virginica_mean = virginica.mean()
print("Virginicar mean\n",virginica_mean)
virginica_std = virginica.std()
print("Virginica Standard Deviation\n",virginica_std)

Virginicar mean
 SL    6.588
SW    2.974
PL    5.552
PW    2.026
dtype: float64
Virginica Standard Deviation
 SL    0.635880
SW    0.322497
PL    0.551895
PW    0.274650
dtype: float64


In [21]:
#conditional probability
mean_values = {
    'setosa': np.array([5.006, 3.428, 1.462, 0.246]),
    'versicolor': np.array([5.936, 2.77, 4.26, 1.326]),
    'virginica': np.array([6.588, 2.974, 5.552, 2.026])     
}

std_values = {
    'setosa': np.array([0.35249, 0.379064, 0.173664, 0.105386]),
    'versicolor': np.array([0.516171, 0.313798, 0.469911, 0.197753]),
    'virginica': np.array([0.635880, 0.322497, 0.551895, 0.27465]),
}
prior_probabilities = {
    'setosa': 1/3,
    'versicolor': 1/3,
    'virginica': 1/3
}

# Given observations
observations = np.array([[4.5, 3.0, 5.6, 2.1], [5.4, 2.6, 4.5, 0.0]])

# Function to calculate conditional probability using Gaussian probability density function
def calculate_conditional_probability(x, mean, std):
    exponent = -((x - mean) ** 2) / (2 * std ** 2)
    return (1 / (math.sqrt(2 * math.pi) * std)) * math.exp(exponent)

#Multiply same class conditional probability
for observe in observations:
    conditional_probs = {}
    posterior_probs = {}

    # Calculate conditional probabilities for each class
    for class_label in mean_values.keys():
        conditional_probs[class_label] = [calculate_conditional_probability(observe[i], mean_values[class_label][i], std_values[class_label][i]) for i in range(4)]

    # Calculate posterior probabilities for each class
    #Multiply prior probability with probability obtained in 3
    for class_label in mean_values.keys():
        posterior_probs[class_label] = prior_probabilities[class_label] * np.prod(conditional_probs[class_label])

    print(f"\nFor {observe}\n")
    for class_label in mean_values.keys():
        print(f"Step 3 Multiply same class conditional probability {class_label}:\n")
        for i, attribute in enumerate(['Sepal Length', 'Sepal Width', 'Petal Length', 'Petal Width']):
            print(f"P({attribute}|{class_label})", conditional_probs[class_label][i])
        print(f"The {class_label} class conditional probability is:", posterior_probs[class_label])
        print("\n")

    # Display final posterior probabilities
    print("Multiply prior probability with probability obtained in ")
    for class_label in mean_values.keys():
        print(f"The {class_label} class probability is: posterior({class_label}) =", posterior_probs[class_label])
    print("\n")


For [4.5 3.  5.6 2.1]

Step 3 Multiply same class conditional probability setosa:

P(Sepal Length|setosa) 0.40392013907615026
P(Sepal Width|setosa) 0.5563719391353582
P(Petal Length|setosa) 1.1875541655714434e-123
P(Petal Width|setosa) 2.355834634483228e-67
The setosa class conditional probability is: 2.0957414363597506e-191


Step 3 Multiply same class conditional probability versicolor:

P(Sepal Length|versicolor) 0.01612394999498657
P(Sepal Width|versicolor) 0.9718588767383594
P(Petal Length|versicolor) 0.014558959141976361
P(Petal Width|versicolor) 0.0009511904237472855
The versicolor class conditional probability is: 7.233545044047977e-08


Step 3 Multiply same class conditional probability virginica:

P(Sepal Length|virginica) 0.0028588559122621183
P(Sepal Width|virginica) 1.2330281405458783
P(Petal Length|virginica) 0.7201302200978542
P(Petal Width|virginica) 1.4007697304508109
The virginica class conditional probability is: 0.0011852822633235968


Multiply prior probability wi

Observation:
##### 1 [4.5 3.  5.6 2.1] we can see that Virginica has highest probability value 

##### 2 [5.4 2.6 4.5 0. ] we can see that Versicolor has highest probability value 

In [22]:
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

# Load the iris dataset
iris = datasets.load_iris()
X = iris.data
y = iris.target

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a Naïve Bayes classifier
classifier = GaussianNB()

# Train the classifier on the training set
classifier.fit(X_train, y_train)

# Use the classifier to predict the species for the given observations
new_observations = [[4.5, 3.0, 5.6, 2.1], [5.4, 2.6, 4.5, 0.0]]
predictions = classifier.predict(new_observations)

# Print the predicted species for each observation
for i, observation in enumerate(new_observations):
    print(f"Observation {i+1}: {iris.target_names[predictions[i]]}")

Observation 1: virginica
Observation 2: versicolor


##### From above output we can see that In observation 1 we get Virginica to have highest probability and in observation 2 we get Versicolor to have highest probability and it matches our manual calculation 

## 3. Consider the data of play tennis or not. Predict the output for the following cases.
### 1. Outlook: Sunny, Temp: Hot, Humidity: High, Wind = Strong
### 2. Outlook = Overcast, Temp = Mild, humidity = High, Wind = Strong
### Verify your answer with sklearn


In [23]:
# Assigning features and label variables
weather=['Sunny','Sunny','Overcast','Rainy','Rainy','Rainy','Overcast','Sunny','Sunny',
'Rainy','Sunny','Overcast','Overcast','Rainy']
outlook = weather
temp=['Hot','Hot','Hot','Mild','Cool','Cool','Cool','Mild','Cool','Mild','Mild','Mild','Hot','Mild']

humidity = ['High', 'High', 'High', 'High', 'Normal', 'Normal', 'Normal', 'High', 'Normal', 'Normal', 'Normal','High', 'Normal', 'High']
wind = ['Weak', 'Strong', 'Weak', 'Weak', 'Weak', 'Strong', 'Strong', 'Weak', 'Weak', 'Weak', 'Strong', 'Strong', 'Weak', 'Strong']
play=['No','No','Yes','Yes','Yes','No','Yes','No','Yes','Yes','Yes','Yes','Yes','No']

tennis_data = pd.DataFrame(zip(outlook, temp, humidity, wind, play), columns = ['Outlook','Temperature', 'Humidity', 'Wind', 'Play Tennis' ])
tennis_data.to_csv("tennis_data.csv", index = False)
tennis_data

Outlook Temperature Humidity    Wind Play Tennis
0      Sunny         Hot     High    Weak          No
1      Sunny         Hot     High  Strong          No
2   Overcast         Hot     High    Weak         Yes
3      Rainy        Mild     High    Weak         Yes
4      Rainy        Cool   Normal    Weak         Yes
5      Rainy        Cool   Normal  Strong          No
6   Overcast        Cool   Normal  Strong         Yes
7      Sunny        Mild     High    Weak          No
8      Sunny        Cool   Normal    Weak         Yes
9      Rainy        Mild   Normal    Weak         Yes
10     Sunny        Mild   Normal  Strong         Yes
11  Overcast        Mild     High  Strong         Yes
12  Overcast         Hot   Normal    Weak         Yes
13     Rainy        Mild     High  Strong          No

### For Outlook : Sunny, Temp: Hot, Humidity: High, Wind = Strong
##### x ’=(Outlook=Sunny, Temperature=Hot, Humidity=High,Wind=Strong)

#### 1. Calculate prior probability for given class labels
#####  From above table we can see that probability for play tennis is P(Yes) = 9/14 and
##### Probability for not play tennis is P(No) = 5/14

#### 2. Calculate conditional probability with each attributes for each class
##### P(Outlook=Sunny|Play=Yes) = 2/9
##### P(Temperature=Hot|Play=Yes) = 2/9
##### P(Humidity=High|Play=Yes) = 3/9
##### P(Wind=Strong|Play=Yes) = 3/9

##### P(Outlook=Sunny|Play=No) = 3/5
##### P(Temperature=Hot|Play==No) = 2/5
##### P(Humidity=High|Play=No) = 4/5
##### P(Wind=Strong|Play=No) = 3/5

#### 3. Multiply same class conditional probability and Multiply that with priori probability 
##### P(Yes|x ’): [P(Sunny|Yes)P(Hot|Yes)P(High|Yes)P(Strong|Yes)]P(Play=Yes)
##### => (2/9)*(2/9)*(3/9)*(3/9)*(9/14) = 0.003527
##### P(No|x ’): [P(Sunny|No) P(Hot|No)P(High|No)P(Strong|No)]P(Play=No)
##### => (3/5)*(2/5)*(4/5)*(3/5)*(5/14) = 0.0421

#### Observation: Given the fact P(Yes| x’) < P(No| x’), we label x’ to be “No” Answer => No, don't Play Tennis


### For Outlook = Overcast, Temp = Mild, humidity = High, Wind = Strong
##### x ’=(Outlook=Overcast, Temperature=Mild, Humidity=High,Wind=Strong)

#### 1. Calculate prior probability for given class labels
#####  From above table we can see that probability for play tennis is P(Yes) = 9/14 and
##### Probability for not play tennis is P(No) = 5/14

#### 2. Calculate conditional probability with each attributes for each class
##### P(Outlook=Overcast|Play=Yes) = 4/9
##### P(Temperature=Mild|Play=Yes) = 4/9
##### P(Humidity=High|Play=Yes) = 3/9
##### P(Wind=Strong|Play=Yes) = 3/9

##### P(Outlook=Overcast|Play=No) = 0/5
##### P(Temperature=Mild|Play==No) = 2/5
##### P(Humidity=High|Play=No) = 4/5
##### P(Wind=Strong|Play=No) = 3/5

#### 3. Multiply same class conditional probability and Multiply that with priori probability 
##### [P(Overcast|Yes)P(Mild|Yes)P(High|Yes)P(Strong|Yes)]P(Play=Yes)
##### => (4/9)*(4/9)*(3/9)*(3/9)*(9/14)=0.01412
##### P(No|x ’): [P(Overcast|No)P(Mild|No)P(High|No)P(Strong|No)]P(Play=No)
##### =>(0/5)*(2/5)*(4/5)*(3/5)*(5/14)=0.000

#### Observation: Given the fact P(Yes| x’) > P(No| x’), we label x’ to be “Yes” Answer => Yes, Play Tennis

In [24]:
#Predicting to play tennis or no
#test data
tennis_data_test=pd.DataFrame([['Sunny','Hot','High','Strong'],['Overcast','Mild','High','Strong']],columns = ['Outlook','Temperature', 'Humidity', 'Wind' ])
tennis_data_test

Outlook Temperature Humidity    Wind
0     Sunny         Hot     High  Strong
1  Overcast        Mild     High  Strong

In [25]:
# Import LabelEncoder
from sklearn.preprocessing import LabelEncoder
#creating labelEncoder
le = LabelEncoder()

# Converting string labels into numbers.
weather_encoded=le.fit_transform(weather)
print("Outlook:", weather_encoded)

temp_encoded=le.fit_transform(temp)
print("Temp:",temp_encoded)

humidity_encoded=le.fit_transform(humidity)
print("Humidity:",humidity_encoded)

wind_encoded=le.fit_transform(wind)
print("Wind:",wind_encoded)

play_encoded=le.fit_transform(play)
print("Play:",play_encoded)

Outlook: [2 2 0 1 1 1 0 2 2 1 2 0 0 1]
Temp: [1 1 1 2 0 0 0 2 0 2 2 2 1 2]
Humidity: [0 0 0 0 1 1 1 0 1 1 1 0 1 0]
Wind: [1 0 1 1 1 0 0 1 1 1 0 0 1 0]
Play: [0 0 1 1 1 0 1 0 1 1 1 1 1 0]


In [26]:
# Converting string labels into numbers.
weather_encoded_test=le.fit_transform(tennis_data_test['Outlook'])
print("Outlook:", weather_encoded_test)

temp_encoded_test=le.fit_transform(tennis_data_test['Temperature'])
print("Temp:",temp_encoded_test)

humidity_encoded_test=le.fit_transform(tennis_data_test['Humidity'])
print("Humidity:",humidity_encoded_test)

wind_encoded_test=le.fit_transform(tennis_data_test['Wind'])
print("Wind:",wind_encoded_test)


Outlook: [1 0]
Temp: [0 1]
Humidity: [0 0]
Wind: [0 0]


In [27]:
# Creating the training data (X_train) dataframe
X_train = pd.DataFrame(zip(weather_encoded, temp_encoded, humidity_encoded, wind_encoded), columns=['outlook', 'temp', 'humidity', 'wind'])
X_train

outlook  temp  humidity  wind
0         2     1         0     1
1         2     1         0     0
2         0     1         0     1
3         1     2         0     1
4         1     0         1     1
5         1     0         1     0
6         0     0         1     0
7         2     2         0     1
8         2     0         1     1
9         1     2         1     1
10        2     2         1     0
11        0     2         0     0
12        0     1         1     1
13        1     2         0     0

#### Outlook: 0 -> Overcast, 1 -> Rainy 2 -> Sunny
#### Temp: 0 -> Cool, 1 -> Mild, 2 -> Hot
#### Humidity: 0 -> High , 1 -> Normal
#### Wind: 0 -> No 1 -> Yes

In [28]:
# Creating the target variable (y_train) dataframe and flattening it to a 1D array
y_train = pd.DataFrame(play_encoded, columns=['play'])
y_train = y_train.values.ravel()
y_train

array([0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0])

In [29]:
# Creating the testing data (X_test) dataframe
X_test = pd.DataFrame(zip(weather_encoded_test, temp_encoded_test, humidity_encoded_test, wind_encoded_test), columns=['outlook', 'temp', 'humidity', 'wind'])
X_test

outlook  temp  humidity  wind
0        1     0         0     0
1        0     1         0     0

In [30]:
#Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB

#Create a Gaussian Classifier
nb_model = GaussianNB()

In [31]:
nb_model.fit(X_train,y_train)

GaussianNB()

In [32]:
y_pred = nb_model.predict(X_test)
y_pred

array([0, 1])

In [33]:
y_pred_df=pd.DataFrame(y_pred,columns=["Play_tennis"])
y_pred_df=y_pred_df.replace({0:'No',1:'Yes'})
y_pred_df

Play_tennis
0          No
1         Yes

#### Observation: 
##### 1. For Outlook : Sunny, Temp: Hot, Humidity: High, Wind = Strong, we can see that answer is no 
##### 2. For Outlook = Overcast, Temp = Mild, humidity = High, Wind = Strong, we can see that answer is Yes 
##### The answers obtained from manually calculation and using sklearn matches. 